crates/generate/src/quickjs.rs

```diff
@@ -0,0 +1,480 @@
+    pub fn error(&self, args: rquickjs::function::Rest<Value<'_>>) -> rquickjs::Result<()> {
+        eprintln!("Error: {}", format_args(&args));
+        Ok(())
+    }
+}
+
+fn resolve_module_path(base_path: &Path, module_path: &str) -> rquickjs::Result<PathBuf> {
+    let candidates = if module_path.starts_with("./") || module_path.starts_with("../") {
+        let target = base_path.join(module_path);
+        vec![
+            target.with_extension("js"),
+            target.with_extension("json"),
+            target.clone(),
+        ]
+    } else {
+        let local_target = base_path.join(module_path);
+        let node_modules_target = Path::new("node_modules").join(module_path);
+
+        vec![
+            local_target.with_extension("js"),
+            local_target.with_extension("json"),
+            local_target.clone(),
+            node_modules_target.with_extension("js"),
+            node_modules_target.with_extension("json"),
+            node_modules_target,
+        ]
+    };
+
+    for candidate in candidates {
+        if let Ok(resolved) = try_resolve_path(&candidate) {
+            return Ok(resolved);
+        }
+    }
+
+    Err(rquickjs::Error::new_from_js_message(
+        "Error",
+        "ModuleNotFound",
+        format!("Module not found: {module_path}"),
+    ))
+}
+
+fn try_resolve_path(path: &Path) -> rquickjs::Result<PathBuf> {
+    let metadata = std::fs::metadata(path).map_err(|_| {
+        rquickjs::Error::new_from_js_message(
+            "Error",
+            "FileNotFound",
+            format!("Path not found: {}", path.display()),
+        )
+    })?;
+
+    if metadata.is_file() {
+        return Ok(path.to_path_buf());
+    }
+
+    if metadata.is_dir() {
+        let index_path = path.join("index.js");
+        if index_path.exists() {
+            return Ok(index_path);
+        }
+    }
+
+    Err(rquickjs::Error::new_from_js_message(
+        "Error",
+        "ResolutionFailed",
+        format!("Cannot resolve: {}", path.display()),
+    ))
+}
+
+#[rquickjs::function]
+#[allow(clippy::needless_pass_by_value)]
+fn require(ctx: Ctx<'_>, module_path: String) -> rquickjs::Result<Value> {
+    let current_module = MODULE_CONTEXT.current()?;
+    let current_dir = if current_module.is_file() {
+        current_module.parent().unwrap_or(Path::new("."))
+    } else {
+        current_module.as_path()
+    };
+
+    let resolved_path = resolve_module_path(current_dir, &module_path)?;
+
+    let contents = load_file(resolved_path.to_string_lossy().to_string())?;
+
+    MODULE_CONTEXT.push(resolved_path.clone());
+
+    let result = load_module_from_content(&ctx, &resolved_path, &contents);
+
+    MODULE_CONTEXT.pop();
+
+    result
+}
+
+fn load_module_from_content<'a>(
+    ctx: &Ctx<'a>,
+    path: &Path,
+    contents: &str,
+) -> rquickjs::Result<Value<'a>> {
+    if path.extension().is_some_and(|ext| ext == "json") {
+        return ctx.eval::<Value, _>(format!("JSON.parse({contents:?})"));
```

Comment:
This is vulnerable to injection.

Reply:
It's no more vulnerable than shelling out to node and executing arbitrary code.